perceval/converters/qiskit_converter.py

```diff
@@ -0,0 +1,216 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from perceval import PredefinedCircuit, Circuit, Processor, P, Source, BasicState
+from perceval.algorithm.norm import *
+from perceval.algorithm.optimize import optimize
+
+import qiskit
+
+# TODO: add typing on library
+
+min_precision_gate = 1e-4
+
+
+def _swap(perm, port_a, port_b):
+    if port_a != port_b:
+        c = perm[port_a]
+        perm[port_a] = perm[port_b]
+        perm[port_b] = c
+
+def _check_heralds(s, idx, heralds):
+    for k, v in heralds.items():
+        if s[k+idx] != v:
+            return False
+    return True
+
+class QiskitConverter:
+
+    def __init__(self, library, source: Source = None):
+        r"""Initialize qiskit to perceval circuit converter.
+
+        :param library: a component library to use for the conversion
+        :param source: the source used as input. Default is a perfect source.
+        """
+        if source is None:
+            source = Source()
+        self._source = source
+        self._lib = library
+
+    def convert(self, qc: qiskit.QuantumCircuit, heralded: bool = None) -> Processor:
+        r"""Convert a qiskit circuit into a perceval.Processor.
+
+        :param qc: quantum-based qiskit circuit
+        :type qc:  qiskit.QuantumCircuit
+        :param heralded: do we use `heralded_cnot` or `post_processed_cnot`, if not set, all cnot but the
+            last are heralded
+        :return: a processor
+        """
+
+        # count the number of cnot to use during the conversion, will give us the number of herald to handle
+        n_cnot = 0
+        for instruction in qc.data:
+            if instruction[0].name == "cx":
+                n_cnot += 1
+
+        cnot_component_heralded = self._lib.catalog["heralded_cnot"]
+        cnot_component_postprocessed = self._lib.catalog["post_processed_cnot"]
+        generic_2mode_component = self._lib.catalog["generic_2mode"]
+        lower_phase_component = PredefinedCircuit(self._lib.Circuit(2) // (0, self._lib.PS(P("phi2"))))
```

Comment:
as the predefinedcircuit should be readonly, shouldn't it be somewhere in a catalog ?